libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtils.kt

```diff
@@ -1799,86 +1878,7 @@ internal object MapboxRouteLineUtils {
-        destinationIcon: Drawable,
-        iconOffset: List<Double>,
-        iconAnchor: IconAnchor,
-        iconPitchAlignment: IconPitchAlignment
-    ): SymbolLayer {
-        if (style.styleLayerExists(RouteLayerConstants.WAYPOINT_LAYER_ID)) {
-            style.removeStyleLayer(RouteLayerConstants.WAYPOINT_LAYER_ID)
-        }
-
-        if (style.getStyleImage(RouteLayerConstants.ORIGIN_MARKER_NAME) != null) {
-            style.removeStyleImage(RouteLayerConstants.ORIGIN_MARKER_NAME)
-        }
-        originIcon.getBitmap().let {
-            style.addImage(RouteLayerConstants.ORIGIN_MARKER_NAME, it)
-        }
-
-        if (style.getStyleImage(RouteLayerConstants.DESTINATION_MARKER_NAME) != null) {
-            style.removeStyleImage(RouteLayerConstants.DESTINATION_MARKER_NAME)
-        }
-        destinationIcon.getBitmap().let {
-            style.addImage(RouteLayerConstants.DESTINATION_MARKER_NAME, it)
-        }
-
-        return SymbolLayer(
-            RouteLayerConstants.WAYPOINT_LAYER_ID,
-            RouteLayerConstants.WAYPOINT_SOURCE_ID
-        )
-            .iconOffset(iconOffset)
-            .iconAnchor(iconAnchor)
-            .iconImage(
-                match {
-                    toString {
-                        get { literal(RouteLayerConstants.WAYPOINT_PROPERTY_KEY) }
-                    }
-                    literal(RouteLayerConstants.WAYPOINT_ORIGIN_VALUE)
-                    stop {
-                        RouteLayerConstants.WAYPOINT_ORIGIN_VALUE
-                        literal(RouteLayerConstants.ORIGIN_MARKER_NAME)
-                    }
-                    stop {
-                        RouteLayerConstants.WAYPOINT_DESTINATION_VALUE
-                        literal(RouteLayerConstants.DESTINATION_MARKER_NAME)
-                    }
-                }
-            ).iconSize(
-                interpolate {
-                    exponential { literal(1.5) }
-                    zoom()
-                    stop {
-                        literal(0.0)
-                        literal(0.6)
-                    }
-                    stop {
-                        literal(10.0)
-                        literal(0.8)
-                    }
-                    stop {
-                        literal(12.0)
-                        literal(1.3)
-                    }
-                    stop {
-                        literal(22.0)
-                        literal(2.8)
-                    }
-                }
-            )
-            .iconPitchAlignment(iconPitchAlignment)
-            .iconAllowOverlap(true)
-            .iconIgnorePlacement(true)
-            .iconKeepUpright(true)
-    }
-
-    internal fun removeLayersAndSources(style: Style) {
-        style.removeStyleSource(RouteLayerConstants.LAYER_GROUP_1_SOURCE_ID)
-        style.removeStyleSource(RouteLayerConstants.LAYER_GROUP_2_SOURCE_ID)
-        style.removeStyleSource(RouteLayerConstants.LAYER_GROUP_3_SOURCE_ID)
```

Comment:
Why don't you need to remove source layer groups anymore?

Reply:
We could follow an identical "on-demand" removal for layers as well, but it doesn't change anything in practice so I opted to skip these changes to keep this PR small. We can still make them in a separate PR just to re-align the code though.

Reply:
Oh, I think I'm getting it now. The algorithm is the following:
1. add sources and layers;
2. remove layers;
3. if sources are compatible with options, just add layers;
4. if sources are incompatible with options, remove sources, add sources, add layers;

Right?

Reply:
Close, more like:
1.  remove layers if they exist;
2. if sources exist and are compatible with options, just add layers;
3. if sources don't exist or are incompatible with options, remove sources, add sources, add layers;